File: neon_tetris.cs
// neon_tetris.cs
using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;

class NeonTetris
{
    const int WIDTH = 10;
    const int HEIGHT = 20;
    static readonly int[][][] SHAPES = new int[][][]
    {
        new int[][] { new int[] {1,1,1,1} },
        new int[][] { new int[] {1,1}, new int[] {1,1} },
        new int[][] { new int[] {0,1,0}, new int[] {1,1,1} },
        new int[][] { new int[] {1,0,0}, new int[] {1,1,1} },
        new int[][] { new int[] {0,0,1}, new int[] {1,1,1} },
        new int[][] { new int[] {0,1,1}, new int[] {1,1,0} },
        new int[][] { new int[] {1,1,0}, new int[] {0,1,1} }
    };
    static readonly ConsoleColor[] COLORS = new ConsoleColor[]
    {
        ConsoleColor.Cyan, ConsoleColor.Yellow, ConsoleColor.Magenta,
        ConsoleColor.Green, ConsoleColor.Red, ConsoleColor.Blue, ConsoleColor.White
    };

    static int[][] board = new int[HEIGHT][];
    static int score = 0, lines = 0, level = 1;
    static int fallTime = 500;
    static DateTime lastFall = DateTime.Now;
    static bool gameOver = false, paused = false;
    static int[][] currentShape;
    static ConsoleColor currentColor;
    static int pieceX, pieceY;
    static int[][] nextShape;
    static ConsoleColor nextColor;

    static void Main()
    {
        Console.CursorVisible = false;
        Console.Title = "Neon Tetris";
        Console.BackgroundColor = ConsoleColor.Black;
        Console.Clear();

        for (int i = 0; i < HEIGHT; i++) board[i] = new int[WIDTH];

        Random rand = new Random();
        nextShape = SHAPES[rand.Next(SHAPES.Length)];
        nextColor = COLORS[rand.Next(COLORS.Length)];
        SpawnPiece(rand);

        Task.Run(() => InputLoop());

        while (!gameOver)
        {
            if (!paused)
            {
                if ((DateTime.Now - lastFall).TotalMilliseconds > fallTime)
                {
                    MovePiece(0, 1);
                    lastFall = DateTime.Now;
                }
            }
            Draw();
            Thread.Sleep(20);
        }
        Console.SetCursorPosition(0, HEIGHT + 4);
        Console.WriteLine("GAME OVER");
        Console.ReadKey();
    }

    static void SpawnPiece(Random rand)
    {
        currentShape = nextShape;
        currentColor = nextColor;
        nextShape = SHAPES[rand.Next(SHAPES.Length)];
        nextColor = COLORS[rand.Next(COLORS.Length)];
        pieceX = WIDTH / 2 - currentShape[0].Length / 2;
        pieceY = 0;
        if (Collision(currentShape, pieceX, pieceY)) gameOver = true;
    }

    static bool Collision(int[][] shape, int x, int y)
    {
        for (int row = 0; row < shape.Length; row++)
            for (int col = 0; col < shape[row].Length; col++)
                if (shape[row][col] != 0)
                {
                    int bx = x + col;
                    int by = y + row;
                    if (bx < 0 || bx >= WIDTH || by >= HEIGHT) return true;
                    if (by >= 0 && board[by][bx] != 0) return true;
                }
        return false;
    }

    static void LockPiece()
    {
        for (int row = 0; row < currentShape.Length; row++)
            for (int col = 0; col < currentShape[row].Length; col++)
                if (currentShape[row][col] != 0)
                {
                    int by = pieceY + row;
                    int bx = pieceX + col;
                    if (by >= 0) board[by][bx] = 1;
                }
        ClearLines();
        Random rand = new Random();
        SpawnPiece(rand);
    }

    static void ClearLines()
    {
        int cleared = 0;
        for (int y = HEIGHT - 1; y >= 0; y--)
        {
            bool full = true;
            for (int x = 0; x < WIDTH; x++) if (board[y][x] == 0) { full = false; break; }
            if (full)
            {
                for (int ny = y; ny > 0; ny--) board[ny] = board[ny - 1];
                board[0] = new int[WIDTH];
                cleared++;
                y++;
            }
        }
        if (cleared > 0)
        {
            lines += cleared;
            score += cleared * 100;
            level = lines / 10 + 1;
            fallTime = Math.Max(100, 500 - (level - 1) * 30);
        }
    }

    static void MovePiece(int dx, int dy)
    {
        int nx = pieceX + dx;
        int ny = pieceY + dy;
        if (!Collision(currentShape, nx, ny))
        {
            pieceX = nx;
            pieceY = ny;
            return;
        }
        if (dy == 1) LockPiece();
    }

    static void Rotate()
    {
        int rows = currentShape.Length;
        int cols = currentShape[0].Length;
        int[][] rotated = new int[cols][];
        for (int i = 0; i < cols; i++) rotated[i] = new int[rows];
        for (int i = 0; i < rows; i++)
            for (int j = 0; j < cols; j++)
                rotated[j][rows - 1 - i] = currentShape[i][j];
        if (!Collision(rotated, pieceX, pieceY))
        {
            currentShape = rotated;
        }
        else
        {
            foreach (int dx in new int[] { -1, 1 })
                if (!Collision(rotated, pieceX + dx, pieceY))
                {
                    pieceX += dx;
                    currentShape = rotated;
                    break;
                }
        }
    }

    static void Drop()
    {
        while (!Collision(currentShape, pieceX, pieceY + 1)) pieceY++;
        LockPiece();
    }

    static void InputLoop()
    {
        while (true)
        {
            var key = Console.ReadKey(true).Key;
            if (key == ConsoleKey.Q) Environment.Exit(0);
            if (key == ConsoleKey.P) paused = !paused;
            if (paused) continue;
            switch (key)
            {
                case ConsoleKey.LeftArrow: MovePiece(-1, 0); break;
                case ConsoleKey.RightArrow: MovePiece(1, 0); break;
                case ConsoleKey.DownArrow: MovePiece(0, 1); break;
                case ConsoleKey.UpArrow: Rotate(); break;
                case ConsoleKey.Spacebar: Drop(); break;
            }
        }
    }

    static void Draw()
    {
        Console.SetCursorPosition(0, 0);
        // Рамка
        Console.Write("┌");
        for (int i = 0; i < WIDTH; i++) Console.Write("─");
        Console.WriteLine("┐");
        for (int y = 0; y < HEIGHT; y++)
        {
            Console.Write("│");
            for (int x = 0; x < WIDTH; x++)
            {
                char ch = ' ';
                ConsoleColor color = ConsoleColor.White;
                if (board[y][x] != 0) { ch = '█'; color = COLORS[(x + y) % COLORS.Length]; }
                // current
                for (int row = 0; row < currentShape.Length; row++)
                    for (int col = 0; col < currentShape[row].Length; col++)
                        if (currentShape[row][col] != 0)
                        {
                            int by = pieceY + row;
                            int bx = pieceX + col;
                            if (by == y && bx == x) { ch = '█'; color = currentColor; }
                        }
                if (ch != ' ')
                {
                    Console.ForegroundColor = color;
                    Console.Write(ch);
                    Console.ResetColor();
                }
                else Console.Write(' ');
            }
            Console.WriteLine("│");
        }
        Console.Write("└");
        for (int i = 0; i < WIDTH; i++) Console.Write("─");
        Console.WriteLine("┘");
        Console.WriteLine($"Score: {score}  Level: {level}  Lines: {lines}");
        if (paused) Console.WriteLine("PAUSED");
        if (gameOver) Console.WriteLine("GAME OVER");
    }
}

File: NeonTetris.java
// NeonTetris.java
import com.googlecode.lanterna.TerminalSize;
import com.googlecode.lanterna.input.KeyStroke;
import com.googlecode.lanterna.input.KeyType;
import com.googlecode.lanterna.screen.Screen;
import com.googlecode.lanterna.screen.TerminalScreen;
import com.googlecode.lanterna.terminal.DefaultTerminalFactory;
import com.googlecode.lanterna.terminal.Terminal;

import java.io.IOException;
import java.util.Random;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

public class NeonTetris {
    private static final int WIDTH = 10;
    private static final int HEIGHT = 20;
    private static final int[][][] SHAPES = {
        {{1,1,1,1}},
        {{1,1},{1,1}},
        {{0,1,0},{1,1,1}},
        {{1,0,0},{1,1,1}},
        {{0,0,1},{1,1,1}},
        {{0,1,1},{1,1,0}},
        {{1,1,0},{0,1,1}}
    };
    private static final String[] COLORS = {"cyan", "yellow", "magenta", "green", "red", "blue", "white"};

    private Screen screen;
    private int[][] board = new int[HEIGHT][WIDTH];
    private int score = 0, lines = 0, level = 1;
    private int fallTime = 500;
    private long lastFall = System.currentTimeMillis();
    private boolean gameOver = false, paused = false;
    private int[][] currentShape;
    private String currentColor;
    private int pieceX, pieceY;
    private int[][] nextShape;
    private String nextColor;
    private Random rand = new Random();

    public NeonTetris() throws IOException {
        Terminal terminal = new DefaultTerminalFactory().createTerminal();
        screen = new TerminalScreen(terminal);
        screen.startScreen();
        screen.setCursorPosition(null);
        screen.clear();
        nextShape = SHAPES[rand.nextInt(SHAPES.length)];
        nextColor = COLORS[rand.nextInt(COLORS.length)];
        spawnPiece();
        Executors.newSingleThreadScheduledExecutor().scheduleAtFixedRate(this::tick, 0, 20, TimeUnit.MILLISECONDS);
        handleInput();
    }

    private void spawnPiece() {
        currentShape = nextShape;
        currentColor = nextColor;
        nextShape = SHAPES[rand.nextInt(SHAPES.length)];
        nextColor = COLORS[rand.nextInt(COLORS.length)];
        pieceX = WIDTH / 2 - currentShape[0].length / 2;
        pieceY = 0;
        if (collision(currentShape, pieceX, pieceY)) gameOver = true;
    }

    private boolean collision(int[][] shape, int x, int y) {
        for (int row = 0; row < shape.length; row++)
            for (int col = 0; col < shape[row].length; col++)
                if (shape[row][col] != 0) {
                    int bx = x + col;
                    int by = y + row;
                    if (bx < 0 || bx >= WIDTH || by >= HEIGHT) return true;
                    if (by >= 0 && board[by][bx] != 0) return true;
                }
        return false;
    }

    private void lockPiece() {
        for (int row = 0; row < currentShape.length; row++)
            for (int col = 0; col < currentShape[row].length; col++)
                if (currentShape[row][col] != 0) {
                    int by = pieceY + row;
                    int bx = pieceX + col;
                    if (by >= 0) board[by][bx] = 1;
                }
        clearLines();
        spawnPiece();
    }

    private void clearLines() {
        int cleared = 0;
        for (int y = HEIGHT - 1; y >= 0; y--) {
            boolean full = true;
            for (int x = 0; x < WIDTH; x++) if (board[y][x] == 0) { full = false; break; }
            if (full) {
                System.arraycopy(board, 0, board, 1, y);
                board[0] = new int[WIDTH];
                cleared++;
                y++;
            }
        }
        if (cleared > 0) {
            lines += cleared;
            score += cleared * 100;
            level = lines / 10 + 1;
            fallTime = Math.max(100, 500 - (level - 1) * 30);
        }
    }

    private void movePiece(int dx, int dy) {
        int nx = pieceX + dx;
        int ny = pieceY + dy;
        if (!collision(currentShape, nx, ny)) {
            pieceX = nx;
            pieceY = ny;
            return;
        }
        if (dy == 1) lockPiece();
    }

    private void rotate() {
        int rows = currentShape.length;
        int cols = currentShape[0].length;
        int[][] rotated = new int[cols][rows];
        for (int i = 0; i < rows; i++)
            for (int j = 0; j < cols; j++)
                rotated[j][rows - 1 - i] = currentShape[i][j];
        if (!collision(rotated, pieceX, pieceY)) {
            currentShape = rotated;
        } else {
            for (int dx : new int[]{-1, 1}) {
                if (!collision(rotated, pieceX + dx, pieceY)) {
                    pieceX += dx;
                    currentShape = rotated;
                    break;
                }
            }
        }
    }

    private void drop() {
        while (!collision(currentShape, pieceX, pieceY + 1)) pieceY++;
        lockPiece();
    }

    private void tick() {
        if (!paused && !gameOver) {
            long now = System.currentTimeMillis();
            if (now - lastFall > fallTime) {
                movePiece(0, 1);
                lastFall = now;
            }
            draw();
        }
    }

    private void draw() {
        try {
            screen.clear();
            // рамка
            for (int y = 0; y < HEIGHT + 2; y++) {
                for (int x = 0; x < WIDTH + 2; x++) {
                    if (y == 0 || y == HEIGHT + 1 || x == 0 || x == WIDTH + 1)
                        screen.setCharacter(x, y, ' ');
                }
            }
            // поле
            for (int y = 0; y < HEIGHT; y++) {
                for (int x = 0; x < WIDTH; x++) {
                    if (board[y][x] != 0) {
                        screen.setCharacter(x + 1, y + 1, '█');
                    }
                }
            }
            // current
            for (int row = 0; row < currentShape.length; row++) {
                for (int col = 0; col < currentShape[row].length; col++) {
                    if (currentShape[row][col] != 0) {
                        int by = pieceY + row + 1;
                        int bx = pieceX + col + 1;
                        if (by >= 1 && by <= HEIGHT)
                            screen.setCharacter(bx, by, '█');
                    }
                }
            }
            // info
            String info = String.format("Score: %d  Level: %d  Lines: %d", score, level, lines);
            for (int i = 0; i < info.length(); i++) {
                screen.setCharacter(WIDTH + 5 + i, 1, info.charAt(i));
            }
            if (paused) {
                String p = "PAUSED";
                for (int i = 0; i < p.length(); i++)
                    screen.setCharacter(WIDTH/2 + i, HEIGHT/2, p.charAt(i));
            }
            if (gameOver) {
                String go = "GAME OVER";
                for (int i = 0; i < go.length(); i++)
                    screen.setCharacter(WIDTH/2 + i, HEIGHT/2, go.charAt(i));
            }
            screen.refresh();
        } catch (IOException e) { e.printStackTrace(); }
    }

    private void handleInput() {
        try {
            while (true) {
                KeyStroke key = screen.pollInput();
                if (key == null) {
                    Thread.sleep(20);
                    continue;
                }
                if (key.getKeyType() == KeyType.Character && key.getCharacter() == 'q') break;
                if (key.getKeyType() == KeyType.Character && key.getCharacter() == 'p') paused = !paused;
                if (paused) continue;
                switch (key.getKeyType()) {
                    case ArrowLeft: movePiece(-1, 0); break;
                    case ArrowRight: movePiece(1, 0); break;
                    case ArrowDown: movePiece(0, 1); break;
                    case ArrowUp: rotate(); break;
                }
                if (key.getKeyType() == KeyType.Character && key.getCharacter() == ' ') drop();
            }
        } catch (Exception e) { e.printStackTrace(); }
        try { screen.stopScreen(); } catch (IOException e) { e.printStackTrace(); }
        System.exit(0);
    }

    public static void main(String[] args) throws IOException {
        new NeonTetris();
    }
}
